packages/next/build/webpack/plugins/next-drop-client-page-plugin.ts

```diff
@@ -1,29 +1,78 @@
 import { Compiler, Plugin } from 'webpack'
-import { extname } from 'path'
+import { STRING_LITERAL_DROP_BUNDLE } from '../../../next-server/lib/constants'
+
+const PLUGIN_NAME = 'DropAmpFirstPagesPlugin'
+
+// Recursively look up the issuer till it ends up at the root
+function findEntryModule(issuer: any): any {
+  if (issuer.issuer) {
+    return findEntryModule(issuer.issuer)
+  }
+
+  return issuer
+}
+
+function handler(parser: any) {
+  function markAsAmpFirst() {
+    const entryModule = findEntryModule(parser.state.module)
+
+    entryModule.buildInfo.NEXT_ampFirst = true
+  }
+
+  parser.hooks.varDeclarationConst
+    .for(STRING_LITERAL_DROP_BUNDLE)
+    .tap(PLUGIN_NAME, markAsAmpFirst)
+
+  parser.hooks.varDeclarationLet
+    .for(STRING_LITERAL_DROP_BUNDLE)
+    .tap(PLUGIN_NAME, markAsAmpFirst)
+
+  parser.hooks.varDeclaration
+    .for(STRING_LITERAL_DROP_BUNDLE)
+    .tap(PLUGIN_NAME, markAsAmpFirst)
+}
 
 // Prevents outputting client pages when they are not needed
 export class DropClientPage implements Plugin {
   ampPages = new Set()
 
   apply(compiler: Compiler) {
-    compiler.hooks.emit.tap('DropClientPage', (compilation) => {
-      Object.keys(compilation.assets).forEach((assetKey) => {
-        const asset = compilation.assets[assetKey]
+    compiler.hooks.compilation.tap(
+      PLUGIN_NAME,
+      (compilation, { normalModuleFactory }) => {
+        normalModuleFactory.hooks.parser
+          .for('javascript/auto')
+          .tap(PLUGIN_NAME, handler)
 
-        if (asset?._value?.includes?.('__NEXT_DROP_CLIENT_FILE__')) {
-          const cleanAssetKey = assetKey.replace(/\\/g, '/')
-          const page = '/' + cleanAssetKey.split('pages/')[1]
-          const pageNoExt = page.split(extname(page))[0]
+        // @ts-ignore Additional property just for Next.js
+        compilation.__NEXTJS = {}
```

Comment:
I usually use a `WeakMap` to associate external private data to other object. That wouldn't require magic property names, and is straightforward to type (`WeakMap<Compilation, { _ampFirstEntryNames: string[] }>`).

But it work this way too.